### PR TITLE
3025391 Clean Up UX

### PR DESCRIPTION
**Pull request by PapaGrande**

### CODE REVIEW

**Comment by MegaChriz** [comment]
- When I select an other destination, the vertical tab 'Source' gets focus. Ideally, the destination tab would keep focus.
  `FeedTypeForm::ajaxCallback()` probably has the answer to get that right:

  ```php
  // Set URL hash so that the correct settings tab is open.
  if (isset($form[$type . '_configuration']['#id'])) {
    $hash = ltrim($form[$type . '_configuration']['#id'], '#');
    $response->addCommand(new SetHashCommand($hash));
  }
  ```

  There's already commented out code for this available. See `MigrationForm::ajaxCallback()`.

- ```php
  // Filter out empty and null plugins.
  if (in_array($id, ['null', 'empty'])) {
    continue;
  }
  ```
  I see you hard-coded filtering out plugins 'null' and 'empty'. What are these plugins and why do they need to be filtered out?

- Martin and I think that the visibility of panes can be better handled via AJAX than using `#states` as you used in the following:

  ```php
  // Toggle visibility of fetcher & parser fields depending on source value.
  if (in_array($type, ['fetcher', 'parser'])) {
    $form[$type . '_wrapper']['id']['#states'] =[
      'visible' => [
        'select[name="source"]' => ['value' => 'url']
      ]
    ];
  }
  ```

  So replace above with an AJAX implementation.

  I think the following piece of code demonstrates how Feeds handles that:

  ```php
  if ($this->pluginHasForm($plugin, 'configuration')) {
    $form_builder = $this->formFactory->createInstance($plugin, 'configuration');

    $plugin_form = $form_builder->buildConfigurationForm([], $plugin_state);
    $form[$type . '_configuration'] = [
      '#type' => 'details',
      '#group' => 'plugin_settings',
      '#title' => $this->t('@type settings', ['@type' => ucfirst($type)]),
    ];
    $form[$type . '_configuration'] += $plugin_form;
  }
  ```

  So something conditional makes the form element appear.

**Comment by MegaChriz** [comment]
```php
// Filter out empty and null migration plugins as those aren't valid feeds
// source or destination.
if (in_array($id, ['null', 'empty'])) {
  continue;
}
```
According to https://www.drupal.org/project/drupal/issues/2857104, it sounds like that the 'null' plugin is a placeholder for a missing Migration plugin.

The D7 version of Feeds had something like that, there it is called "FeedsMissingPlugin". The D7 version of Feeds fixed filtering out this plugin by overriding the method `pluginType()` and returning "missing" there.

I think we can't use the same strategy here, but what we can do is altering the plugin metadata and add something like a key “hidden” to it. The user interface then skips migration plugins which have “hidden” set to true. This avoids having a hard-coded list of plugins to filter out within the UI code.

**Comment by MegaChriz** [comment]
Perhaps instead of "hidden", “no_ui” would be better for this key, as we already have such one for field types.

**Comment by irinaz** [comment]
Hi, 
I installed branch [PapaGrande/feeds_migrate](/PapaGrande/feeds_migrate) on a new Drupal 8 site  installed it on  new Drupal 8 site  and I am missing button "add migration" in default migration group http://dev-nicu-database.pantheonsite.io/admin/structure/migrate/manage/default/migrations.   Let me know if I need to provide more details. thanks!!

File: modules/feeds_migrate_ui/src/FeedsMigrateUiParserSuggestion.php [hunk 1]
```diff
@@ -34,6 +34,9 @@ public function __construct(DataParserPluginManager $parser_manager) {
     $this->fetcherManager = \Drupal::service('plugin.manager.migrate_plus.data_fetcher');
   }
 
+  /**
+   *
```
Comment (heddn) [inline modules/feeds_migrate_ui/src/FeedsMigrateUiParserSuggestion.php:38]: Missing comment text?

File: modules/feeds_migrate_ui/src/FeedsMigrateUiParserSuggestion.php [hunk 2]
```diff
@@ -67,6 +73,9 @@ public function getSuggestedSelectors($fetcher_plugin_id, $url) {
     }
   }
 
+  /**
+   *
```
Comment (heddn) [inline modules/feeds_migrate_ui/src/FeedsMigrateUiParserSuggestion.php:77]: Missing comment text?

File: modules/feeds_migrate_ui/src/FeedsMigrateUiParserSuggestion.php [hunk 3]
```diff
@@ -49,6 +52,9 @@ public function getSuggestedParser($data) {
     return FALSE;
   }
 
+  /**
+   *
```
Comment (heddn) [inline modules/feeds_migrate_ui/src/FeedsMigrateUiParserSuggestion.php:56]: Missing comment text?

File: modules/feeds_migrate_ui/src/FeedsMigrateUiParserSuggestion.php [hunk 4]
```diff
@@ -85,6 +94,9 @@ protected function getSuggestedSelectorsXml($contents, &$suggestions = [], $path
     return $suggestions;
   }
 
+  /**
+   *
```
Comment (heddn) [inline modules/feeds_migrate_ui/src/FeedsMigrateUiParserSuggestion.php:98]: Missing comment text?

File: modules/feeds_migrate_ui/src/Form/MigrationMappingForm.php [hunk 5]
```diff
@@ -400,7 +401,7 @@ public function save(array $form, FormStateInterface $form_state) {
     $this->messenger()->AddMessage($this->t('Migration mapping for migration 
         @migration has been updated.', [
           '@migration' => $migration->label(),
-    ]));
+        ]));
```
Comment (heddn) [inline modules/feeds_migrate_ui/src/Form/MigrationMappingForm.php:404]: whitespace.

File: modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php [hunk 6]
```diff
@@ -342,15 +341,15 @@ public function save(array $form, FormStateInterface $form_state) {
       // If we edited an existing mapping.
       $this->messenger()->addMessage($this->t('Migration mapping for field 
         @destination_field has been updated.', [
-        '@destination_field' => $this->migrationEntityHelper()->getMappingFieldLabel($this->key)
-      ]));
+          '@destination_field' => $this->migrationEntityHelper()->getMappingFieldLabel($this->key),
```
Comment (heddn) [inline modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php:344]: Whitespace.

File: modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php [hunk 7]
```diff
@@ -342,15 +341,15 @@ public function save(array $form, FormStateInterface $form_state) {
...
+        ]));
     }
     else {
       // If we created a new mapping.
       $this->messenger()->addMessage($this->t('Migration mapping for field
         @destination_field has been added.', [
-        '@destination_field' => $this->migrationEntityHelper()->getMappingFieldLabel($this->key)
-      ]));
+          '@destination_field' => $this->migrationEntityHelper()->getMappingFieldLabel($this->key),
+        ]));
```
Comment (heddn) [inline modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php:352]: Whitespace.

File: modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php [hunk 8]
```diff
@@ -370,16 +369,17 @@ public function ajaxCallback(array $form, FormStateInterface $form_state) {
     return $form;
   }
 
-  /** ----------- */
+  /**
+   * ----------- .*/
```
Comment (heddn) [inline modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php:373]: Missing comment text?

**Review by heddn — CHANGES_REQUESTED** [review]

**Pull request closed without merging**